src/common/state/redis.c

```diff
@@ -820,144 +762,158 @@ void redis_task_table_get_task_callback(redisAsyncContext *c,
+  /* Call the done callback if there is one. */
+  if (callback_data->done_callback != NULL) {
+    task_table_done_callback done_callback = callback_data->done_callback;
+    done_callback(callback_data->id, callback_data->user_context);
   }
-}
-
-void redis_task_table_add_task(table_callback_data *callback_data) {
-  redis_task_table_publish(callback_data, false);
+  /* Clean up the timer and callback. */
+  destroy_timer_callback(db->loop, callback_data);
 }
 
 void redis_task_table_update(table_callback_data *callback_data) {
-  redis_task_table_publish(callback_data, true);
-}
-
-void redis_task_table_publish_push_callback(redisAsyncContext *c,
-                                            void *r,
-                                            void *privdata) {
-  LOG_DEBUG("Calling publish push callback");
-  REDIS_CALLBACK_HEADER(db, callback_data, r);
-  CHECK(callback_data->requests_info != NULL);
-  ((bool *) callback_data->requests_info)[PUBLISH_PUSH_INDEX] = true;
+  db_handle *db = callback_data->db_handle;
+  task *task = callback_data->data;
+  task_id task_id = task_task_id(task);
+  db_client_id local_scheduler_id = task_local_scheduler(task);
+  scheduling_state state = task_state(task);
 
-  if (((bool *) callback_data->requests_info)[PUBLISH_PUBLISH_INDEX] == true) {
-    if (callback_data->done_callback) {
-      task_table_done_callback done_callback = callback_data->done_callback;
-      done_callback(callback_data->id, callback_data->user_context);
-    }
-    destroy_timer_callback(db->loop, callback_data);
+  CHECKM(task != NULL, "NULL task passed to redis_task_table_update.");
+  int status = redisAsyncCommand(
+      db->context, redis_task_table_update_callback,
+      (void *) callback_data->timer_id, "RAY.TASK_TABLE_UPDATE %b %d %b",
+      task_id.id, sizeof(task_id.id), state, local_scheduler_id.id,
+      sizeof(local_scheduler_id.id));
+  if ((status == REDIS_ERR) || db->context->err) {
+    LOG_REDIS_DEBUG(db->context, "error in redis_task_table_update");
   }
 }
 
-void redis_task_table_publish_publish_callback(redisAsyncContext *c,
-                                               void *r,
-                                               void *privdata) {
-  LOG_DEBUG("Calling publish publish callback");
-  REDIS_CALLBACK_HEADER(db, callback_data, r);
-  CHECK(callback_data->requests_info != NULL);
-  ((bool *) callback_data->requests_info)[PUBLISH_PUBLISH_INDEX] = true;
-
-  if (((bool *) callback_data->requests_info)[PUBLISH_PUSH_INDEX] == true) {
-    if (callback_data->done_callback) {
-      task_table_done_callback done_callback = callback_data->done_callback;
-      done_callback(callback_data->id, callback_data->user_context);
-    }
-    destroy_timer_callback(db->loop, callback_data);
-  }
+/* The format of the payload is described in ray_redis_module.c and is
+ * "<task ID> <state> <local scheduler ID> <task specification>". */
+void parse_task_table_subscribe_callback(char *payload,
```

Comment:
Good point, unfortunately sscanf cannot scan strings that contain "\0" at the moment, which is rather unfortunate. We might want to roll our own simple implementation in the future, but that's for later ;)